In [1]:
import time
import pandas as pd
import altair as alt
import numpy as np
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split, GridSearchCV, RandomizedSearchCV
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.preprocessing import OneHotEncoder, PowerTransformer
from sklearn.metrics import mean_squared_error, mean_absolute_error

from sklearn.model_selection import RandomizedSearchCV
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder

from xgboost import XGBRegressor
from sklearn.ensemble import RandomForestRegressor, RandomForestClassifier
from sklearn.preprocessing import StandardScaler


#Download libraries

alt.data_transformers.disable_max_rows()

# import other functions
from imputer import *
from feature_eng import *
from drop import *

In [2]:
alt.data_transformers.disable_max_rows()

DataTransformerRegistry.enable('default')

In [3]:
# defining function to report train and test error

def report_performance(model, X_train, y_train, X_valid, y_valid, 
                       mode='mean'):
    """
    Evaluate train and validation performance on a fitted model.
    
    Parameters
    ---------     
    model: sklearn.ensemble._gb.GradientBoostingRegressor
        scikit-learn model
    X_train: pandas.core.frame.DataFrame
        X of training set
    y_train: pandas.core.series.Series
        y of training set
    X_valid: pandas.core.frame.DataFrame        
        X of validation set
    y_valid: pandas.core.series.Series
        y of validation set     
    mode: string
        'mean' or 'median'
    
    Returns
    -------
    errors: list
        
    """
    if mode == 'mean':
        errors = [(mean_squared_error(y_train, model.predict(X_train)))**0.5, 
                  (mean_squared_error(y_valid, model.predict(X_valid)))**0.5]
        metric = 'squared'
        
    elif mode == 'median':
        errors = [mean_absolute_error(y_train, model.predict(X_train)), 
                  mean_absolute_error(y_valid, model.predict(X_valid))]
        metric= 'absolute'
    
    print('Training mean', metric, 'error:', errors[0])
    print('Validation mean', metric, 'error:', errors[1])

In [4]:
# defining a function to plot error distribution

def plot_resid(model, X_train=None, y_train=None, X_valid=None, y_valid=None, plot = 'both'):
    d = dict()
    
    if plot != 'valid':
        train_df = pd.DataFrame({'Predicted Train':model.predict(X_train), 'True Train':y_train})
        train_df['Train Error Distance'] =  train_df['Predicted Train'] - train_df['True Train']
        train_df['Train Error Proportion'] =  train_df['Predicted Train']/train_df['True Train']
        train_dist = alt.Chart(train_df).mark_circle().encode(alt.X("True Train:Q"), y=alt.Y('Train Error Distance:Q'))
        train_prop = alt.Chart(train_df).mark_circle().encode(alt.X("True Train:Q"), y=alt.Y('Train Error Proportion:Q'))
        d["Train_Distance"] = train_dist
        d["Train_Proportion"] = train_prop
    else:
        d["Valid_Distance"] = "No training set inputted"
        d["Valid_Proportion"] = "No training set inputted"
        
    if plot != 'train':
        valid_df = pd.DataFrame({'Predicted Valid':model.predict(X_valid), 'True Valid':y_valid})
        valid_df['Valid Error Distance'] =  valid_df['Predicted Valid'] - valid_df['True Valid']
        valid_df['Valid Error Proportion'] =  valid_df['Predicted Valid']/valid_df['True Valid']
        
        valid_dist = alt.Chart(valid_df).mark_circle().encode(alt.X("True Valid:Q"), y=alt.Y('Valid Error Distance:Q'))
        valid_prop = alt.Chart(valid_df).mark_circle().encode(alt.X("True Validation:Q"), y=alt.Y('Validation Error Proportion:Q'))
        d["Valid_Distance"] = valid_dist
        d["Valid_Proportion"] = valid_prop
    else:
        d["Valid_Distance"] = "No validation set inputted"
        d["Valid_Proportion"] = "No validation set inputted"
        
    return d

In [5]:
df = pd.read_csv('../data/train_data.zip')

In [6]:
df.head()

external_id  month  year  monthly_number_of_sessions  \
0     1804425      8  2018                           0   
1     1812706      2  2019                           0   
2     1812706      3  2019                           0   
3     1812706     11  2018                           0   
4     1812706      9  2018                           0   

   monthly_unique_sessions  monthly_repeated_sessions  \
0                        0                          0   
1                        0                          0   
2                        0                          0   
3                        0                          0   
4                        0                          0   

   monthly_avg_length_of_session  monthly_avg_light_activity  \
0                            0.0                         0.0   
1                            0.0                         0.0   
2                            0.0                         0.0   
3                            0.0                         0.0   
4                            0.0                         0.0   

   monthly_avg_moderate_activity  monthly_avg_vigorous_activity  ...  \
0                            0.0                            0.0  ...   
1                            0.0                            0.0  ...   
2                            0.0                            0.0  ...   
3                            0.0                            0.0  ...   
4                            0.0                            0.0  ...   

   avg_wind_9_10  avg_wind_10_11  avg_wind_11_12  avg_wind_12_above  \
0            0.0             0.0             0.0                0.0   
1            0.0             0.0             0.0                0.0   
2            0.0             0.0             0.0                0.0   
3            0.0             0.0             0.0                0.0   
4            0.0             0.0             0.0                0.0   

   perfect_days  unacast_session_count     hpi  \
0           0.0                   90.0  244.20   
1           4.0                   27.0  258.95   
2           4.0                   27.0  258.95   
3           3.0                   24.0  258.95   
4           0.0                   12.0  258.95   

   state_and_local_amount_per_capita  state_amount_per_capita  \
0                           0.157475                 0.009783   
1                           0.157475                 0.009783   
2                           0.157475                 0.009783   
3                           0.157475                 0.009783   
4                           0.157475                 0.009783   

   local_amount_per_capita  
0                 0.147692  
1                 0.147692  
2                 0.147692  
3                 0.147692  
4                 0.147692  

[5 rows x 861 columns]

### Running preprocessing and data split

In [7]:
# drop rows missing target variable
df = drop_missing_unacast(df)
# dropping canadian playground
df = df.query("external_id != 'CA00070678'")
# create X and y
X = df.drop('unacast_session_count', axis=1)
y = df.loc[:, 'unacast_session_count']
# split the data
X_train, X_valid, y_train, y_valid = train_test_split(X, y, 
                                                    test_size=0.2,
                                                      random_state=2020)
# impute NaN values
result = impute_data(X_train, X_valid)
X_train = result[0]
X_valid = result[1] 
# perform feature eng
X_train = comb_cols(X_train)
X_valid = comb_cols(X_valid)
# perform dropping
X_train = drop_columns(X_train)
X_valid = drop_columns(X_valid)

In [8]:
# preprocess categoricla columns (OHE)
X_train_valid = clean_categorical(X_train, X_valid)
X_train = X_train_valid[0]
X_valid = X_train_valid[1]

In [9]:
# double check no NaN's left in data
y_train.isna().sum().sum()

0

### Running single model fit to test

In [19]:
# model 1
rf = RandomForestRegressor(n_estimators=1500, max_depth = 15, min_samples_split = 100, max_features = 0.7,n_jobs=-1, bootstrap = True)

In [20]:
t0 = time.time()
rf.fit(X_train, y_train)
t1 = time.time()
tr_time = t1-t0

In [21]:
# model 1
tr_time/60

16.52150979042053

In [22]:
# model 1 RMSE
report_performance(rf, X_train, y_train, X_valid, y_valid)

Training mean squared error: 138.82634321074957
Validation mean squared error: 151.0121505087009


In [23]:
# model 1 R squared 
rf.score(X_train,y_train)

0.7368927783281327

In [24]:
# model 1 R squared 
rf.score( X_valid, y_valid)

0.6797252667872987

### Top 10 most important features in the model

ranked based on impurity metric, which is a measures to dissimilarity within groups.

[sklearn page](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestRegressor.html)

[Gini Impurity](https://en.wikipedia.org/wiki/Decision_tree_learning#Gini_impurity), under `3.1 Gini Impurity`

In [25]:
# model 1 most important features
importances = rf.feature_importances_
n = 10
indices = (-importances).argsort()[:n]
X_train.iloc[:, list(indices)].columns.to_list()

['houses_per_sq_km',
 'B08301e10',
 'walk_score',
 'Republicans_2016',
 'B17020e6',
 'state_amount_per_capita',
 'Libertarians_2016',
 'B25012e17',
 'B08301e6',
 'B19101e8']

### setting hyperparameter dictionary for optimization

In [8]:
# parameters for grid search
max_depth = [3,20,15,5,8]
min_samples_leaf = [50,125,200]
max_features = [0.1*i for i in range(1,8)]
bootstrap = True
#max_samples = [0.05*i for i in range(14,21)]
d = {"max_depth":max_depth, "min_samples_leaf":min_samples_leaf, "max_features":max_features}

In [11]:
# size of grid
c = 1
for val in d.values():
    c *= len(val)
c

105

In [6]:
d.values()

dict_values([[3, 5, 7], [50, 125, 200], [0.1, 0.2, 0.30000000000000004, 0.4, 0.5, 0.6000000000000001, 0.7000000000000001]])

### Running 3 Randomized grid search for optimization

In [12]:
# define grid search
t0 = time.time()
rf_cv = RandomForestRegressor(n_estimators=500, bootstrap = True,n_jobs=-1)
rgscv = RandomizedSearchCV(rf_cv,n_iter=10,param_distributions=d,return_train_score=True,n_jobs=-1  ,scoring=['neg_root_mean_squared_error'], refit=False)
search = rgscv.fit(X_train, y_train)
t1 = time.time()
cv_time = t1-t0

In [ ]:
cv_time/60/60

In [14]:
# grid search results
d = search.cv_results_

pd.DataFrame(data=d)#.sort_values(by='rank_test_neg_root_mean_squared_error	')

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0    1606.684599      2.491543         1.162499        0.156037   
1     144.309173      1.215526         0.608651        0.043203   
2     176.695714      0.767955         0.827943        0.080414   
3     892.502714      8.388668         0.878335        0.013495   
4     426.849827      2.661796         0.564516        0.052117   
5    3776.644686    517.590457         3.688427        0.385812   
6    2840.053057    338.792729         3.801004        2.157962   
7    1778.256419     11.925694         4.335996        1.653517   
8     868.253258     45.537747         2.392829        0.893235   
9     829.860981    123.569716         1.621974        1.458029   

  param_min_samples_leaf param_max_features param_max_depth  \
0                     50                0.5              20   
1                    125                0.2               3   
2                    125                0.1               8   
3                     50                0.5               8   
4                     50                0.6               3   
5                    200                0.7               8   
6                    125                0.3              15   
7                    200                0.2              20   
8                     50                0.3               3   
9                     50                0.1              20   

                                              params  \
0  {'min_samples_leaf': 50, 'max_features': 0.5, ...   
1  {'min_samples_leaf': 125, 'max_features': 0.2,...   
2  {'min_samples_leaf': 125, 'max_features': 0.1,...   
3  {'min_samples_leaf': 50, 'max_features': 0.5, ...   
4  {'min_samples_leaf': 50, 'max_features': 0.600...   
5  {'min_samples_leaf': 200, 'max_features': 0.70...   
6  {'min_samples_leaf': 125, 'max_features': 0.30...   
7  {'min_samples_leaf': 200, 'max_features': 0.2,...   
8  {'min_samples_leaf': 50, 'max_features': 0.300...   
9  {'min_samples_leaf': 50, 'max_features': 0.1, ...   

   split0_test_neg_root_mean_squared_error  \
0                              -146.284670   
1                              -193.719139   
2                              -182.425875   
3                              -158.126367   
4                              -182.441978   
5                              -183.970463   
6                              -172.883108   
7                              -185.611910   
8                              -184.241727   
9                              -152.073831   

   split1_test_neg_root_mean_squared_error  ...  \
0                              -152.649436  ...   
1                              -196.625389  ...   
2                              -185.751417  ...   
3                              -163.919682  ...   
4                              -185.942054  ...   
5                              -188.792381  ...   
6                              -178.030591  ...   
7                              -188.551620  ...   
8                              -185.755599  ...   
9                              -157.344081  ...   

   mean_test_neg_root_mean_squared_error  \
0                            -155.746717   
1                            -201.431892   
2                            -190.263214   
3                            -166.536875   
4                            -189.691303   
5                            -193.035813   
6                            -181.339516   
7                            -193.369914   
8                            -190.558440   
9                            -160.830755   

   std_test_neg_root_mean_squared_error  \
0                              7.935287   
1                              6.330096   
2                              6.719719   
3                              7.413346   
4                              6.841661   
5                              6.562291   
6                              6.449368   
7                              6.474536   
8            

In [11]:
# second grid search
t0 = time.time()
rf_cv = RandomForestRegressor(n_estimators=500, bootstrap = True,n_jobs=-1)
rgscv1 = RandomizedSearchCV(rf_cv,n_iter=10,param_distributions=d,return_train_score=True,n_jobs=-1  ,scoring=['neg_root_mean_squared_error'], refit=False)
search1 = rgscv1.fit(X_train, y_train)
t1 = time.time()
cv_time = t1-t0
d1 = search1.cv_results_

In [12]:
pd.DataFrame(data=d1)

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0    5536.621997     19.954894         3.499835        0.623025   
1    5067.753646     61.576164         3.504749        0.201787   
2    2395.255626      8.373793         3.383983        0.244566   
3    2058.006819     18.018362         3.351321        0.337233   
4    1867.865213     45.371774         2.927546        1.180799   
5    4379.777742     53.856829         2.848914        1.549910   
6    4166.188082    938.182500         2.002912        1.789938   
7     488.751607     24.033742         2.865140        1.120160   
8     605.040898     10.290444         3.517885        0.297434   
9     245.913585     56.319772         1.752811        0.957482   

  param_min_samples_leaf param_max_features param_max_depth  \
0                    200                0.6              15   
1                    125                0.7               8   
2                    125                0.5               5   
3                    125                0.7               3   
4                    200                0.4               5   
5                    125                0.6               8   
6                    200                0.7               8   
7                     50                0.1               5   
8                     50                0.2               3   
9                     50                0.1               3   

                                              params  \
0  {'min_samples_leaf': 200, 'max_features': 0.60...   
1  {'min_samples_leaf': 125, 'max_features': 0.70...   
2  {'min_samples_leaf': 125, 'max_features': 0.5,...   
3  {'min_samples_leaf': 125, 'max_features': 0.70...   
4  {'min_samples_leaf': 200, 'max_features': 0.4,...   
5  {'min_samples_leaf': 125, 'max_features': 0.60...   
6  {'min_samples_leaf': 200, 'max_features': 0.70...   
7  {'min_samples_leaf': 50, 'max_features': 0.1, ...   
8  {'min_samples_leaf': 50, 'max_features': 0.2, ...   
9  {'min_samples_leaf': 50, 'max_features': 0.1, ...   

   split0_test_neg_root_mean_squared_error  \
0                              -181.454794   
1                              -177.972810   
2                              -183.868052   
3                              -192.714838   
4                              -188.343625   
5                              -177.707792   
6                              -183.760395   
7                              -176.466246   
8                              -187.572379   
9                              -194.009646   

   split1_test_neg_root_mean_squared_error  ...  \
0                              -186.234470  ...   
1                              -183.164258  ...   
2                              -189.378081  ...   
3                              -197.400518  ...   
4                              -193.483080  ...   
5                              -183.163354  ...   
6                              -188.837142  ...   
7                              -179.817985  ...   
8                              -187.974343  ...   
9                              -192.425506  ...   

   mean_test_neg_root_mean_squared_error  \
0                            -190.461063   
1                            -186.550786   
2                            -192.352498   
3                            -200.799172   
4                            -197.437788   
5                            -186.193732   
6                            -192.897731   
7                            -183.642707   
8                            -193.270796   
9                            -199.236771   

   std_test_neg_root_mean_squared_error  \
0                              6.624129   
1                              6.384466   
2                              6.249743   
3                              5.950320   
4                              6.696089   
5                              6.336114   
6                              6.551268   
7                              6.974695   
8            

In [10]:
# third grid search
t0 = time.time()
rf_cv = RandomForestRegressor(n_estimators=1000, bootstrap = True,n_jobs=-1)
rgscv2 = RandomizedSearchCV(rf_cv,n_iter=5,param_distributions=d,return_train_score=True,n_jobs=-1  ,scoring=['neg_root_mean_squared_error'], refit=False)
search2 = rgscv2.fit(X_train, y_train)
t1 = time.time()
cv_time = t1-t0
d2 = search2.cv_results_

In [11]:
pd.DataFrame(data=d2)

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0     856.498512      0.403155         0.679534        0.037446   
1    1149.252890      1.294673         0.902257        0.095812   
2    1375.778820      0.896225         0.838490        0.021730   
3     979.230218     23.538915         0.753645        0.110153   
4     703.786022    216.524648         0.590293        0.166373   

  param_min_samples_leaf param_max_features param_max_depth  \
0                     50                0.6               3   
1                    125                0.5               5   
2                    125                0.6               5   
3                     50                0.7               3   
4                    200                0.4               5   

                                              params  \
0  {'min_samples_leaf': 50, 'max_features': 0.600...   
1  {'min_samples_leaf': 125, 'max_features': 0.5,...   
2  {'min_samples_leaf': 125, 'max_features': 0.60...   
3  {'min_samples_leaf': 50, 'max_features': 0.700...   
4  {'min_samples_leaf': 200, 'max_features': 0.4,...   

   split0_test_neg_root_mean_squared_error  \
0                              -182.113708   
1                              -184.274431   
2                              -184.770366   
3                              -182.358169   
4                              -188.642619   

   split1_test_neg_root_mean_squared_error  ...  \
0                              -185.856621  ...   
1                              -189.193007  ...   
2                              -190.235264  ...   
3                              -187.283885  ...   
4                              -193.326993  ...   

   mean_test_neg_root_mean_squared_error  \
0                            -189.552145   
1                            -192.421129   
2                            -193.050712   
3                            -190.321842   
4                            -197.530619   

   std_test_neg_root_mean_squared_error  \
0                              6.906814   
1                              6.140397   
2                              6.099773   
3                              6.904825   
4                              6.607228   

   rank_test_neg_root_mean_squared_error  \
0                                      1   
1                                      3   
2                                      4   
3                                      2   
4                                      5   

   split0_train_neg_root_mean_squared_error  \
0                               -189.790290   
1                               -192.711922   
2                               -193.541250   
3                               -190.311495   
4                               -197.282939   

   split1_train_neg_root_mean_squared_error  \
0                               -188.133360   
1                               -191.822793   
2                               -192.958607   
3                               -189.216369   
4                               -196.730750   

   split2_train_neg_root_mean_squared_error  \
0                               -184.785702   
1                               -187.493221   
2                               -188.259516   
3                               -185.558852   
4                               -194.155290   

   split3_train_neg_root_mean_squared_error  \
0                               -189.808338   
1                               -191.539452   
2                               -192.088629   
3                               -190.376237   
4                               -197.277656   

   split4_train_neg_root_mean_squared_error  \
0                               -188.100206   
1                               -191.769016   
2                               -191.916753   
3                               -188.394686   
4                               -196.980327   

   mean_train_neg_root_mean_squared_error  \
0                             -188.123579   
1     

### Additional single model runs to test additional combinations

In [12]:
# model 1
rf = RandomForestRegressor(n_estimators=1500, max_depth = 20, min_samples_split = 50, max_features = 0.5,n_jobs=-1, bootstrap = True)
t0 = time.time()
rf.fit(X_train, y_train)
t1 = time.time()
tr_time = t1-t0

In [13]:
# model 1
report_performance(rf, X_train, y_train, X_valid, y_valid)

Training mean squared error: 122.616826254174
Validation mean squared error: 139.73050716408426


In [14]:
# model 2
rf = RandomForestRegressor(n_estimators=1500, max_depth = 30, min_samples_split = 25, max_features = 0.5,n_jobs=-1, bootstrap = True)
t0 = time.time()
rf.fit(X_train, y_train)
t1 = time.time()
tr_time = t1-t0

In [15]:
# model 2
report_performance(rf, X_train, y_train, X_valid, y_valid)

Training mean squared error: 101.2009516264235
Validation mean squared error: 129.74548851998952


In [10]:
# model 3
rf = RandomForestRegressor(n_estimators=1500, max_depth = 25, min_samples_split = 25, max_features = 0.5,n_jobs=-1, bootstrap = True)
t0 = time.time()
rf.fit(X_train, y_train)
t1 = time.time()
tr_time = t1-t0

In [13]:
# model 3
tr_time/60

17.91033939123154

In [28]:
# model 3
report_performance(rf, X_train, y_train, X_valid, y_valid)

Training mean squared error: 102.73709137777978
Validation mean squared error: 130.4186745486596


In [53]:
p = plot_resid(rf, X_train, y_train, X_valid, y_valid)

In [30]:
p

{'Train_Distance': alt.Chart(...),
 'Train_Proportion': alt.Chart(...),
 'Valid_Distance': alt.Chart(...),
 'Valid_Proportion': alt.Chart(...)}

### Residuals plot for validation set

In [54]:
# model 3
p['Valid_Distance']

alt.Chart(...)

In [38]:

valid_df = pd.DataFrame({'Predicted Valid':rf.predict(X_valid), 'True Valid':y_valid})
valid_df['Valid Error Distance'] =  valid_df['Predicted Valid'] - valid_df['True Valid']
valid_df['Valid Error Proportion'] =  valid_df['Predicted Valid']/valid_df['True Valid']
valid_df

Predicted Valid  True Valid  Valid Error Distance  \
12721       251.854696       387.0           -135.145304   
17871        60.851852        56.0              4.851852   
46441        46.475018        31.0             15.475018   
48833       122.663007       131.0             -8.336993   
50069        96.940607        32.0             64.940607   
...                ...         ...                   ...   
36902        75.937002        62.0             13.937002   
5882         56.639207        71.0            -14.360793   
22158        26.966143        29.0             -2.033857   
35672       109.599462       156.0            -46.400538   
31331        30.388826        27.0              3.388826   

       Valid Error Proportion  
12721                0.650787  
17871                1.086640  
46441                1.499194  
48833                0.936359  
50069                3.029394  
...                       ...  
36902                1.224790  
5882                 0.797735  
22158                0.929867  
35672                0.702561  
31331                1.125512  

[9898 rows x 4 columns]

In [60]:
len(y_train)

39592

In [58]:
len(np.unique(valid_df['Predicted Valid']))

9898

In [59]:
len(np.unique(rf.predict(X_train)))

39592

In [41]:
alt.Chart(valid_df).mark_circle().encode(alt.X("True Valid:Q"), y=alt.Y('Valid Error Distance:Q'))


alt.Chart(...)

## Summary

- Second round of modeling with SKlearns RandomForestRegressor to minimize the RMSE (finding the mean)
    - MAE minimization in SKlearns random forest is problmatic due to implamentation, which leads to **very** long run times

- Input data
    - `playground_stats_capped.csv`
    - `unacast_session_count` was capped at 4000 because the initial round of modeling revealed that outliers were inflating the RMSE values
    - Updated preprocessing functions
   
- Tuning and optimization
    - Assuming that the quality of the random forest increases linearly with the number of trees, a medium-ish number was selected to save time (n=500 or 1000)
    - optomized over `max_depth`, `min_samples_leaf`, and `max_features`. Explnation on each can be found [here](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestRegressor.html), used a wider range of possible values then in perliminary modeling.
    - best train RMSE was generally around 100-120, and validation minimized around 130, seems to overfit slightly less then in the perliminary modeling
    - the best run included the following hyper parameters:
        - `max_depth = 20`
        - `min_samples_split = 50`
        - `max_features = 0.5`
    - `max_depth` was higher then anticipated, and the model continued to perform better with a high-ish `max_features`
    
- Predictions
    - Due to structure of tree regressors, the end prediction is the average value of the training targets in the corresponding leaf (final node), however currently the model does not predict 0, probably due to how the target variable is currently imputed
      

**Main takeaways**
- The drop in RMSE values is mostly attributable to the capping of the target variable
- Hit a wall: difficult to get validation RMSE to dip below 100
